python-lib/text_extraction/__init__.py

```diff
@@ -54,4 +54,108 @@ def extract_text_content(file_bytes, extension, with_pandoc):
+    header_stack = []
+    initial_metadata = {}
+
+    for line in lines:
+        stripped_line = line.strip()
+
+        # Check each line against each of the header types (e.g., #, ##), header_level is the number of '#'
+        for header_level in range(1, 4):
+            # Check if line starts with a header that we intend to split on
+            if stripped_line.startswith("#" * header_level) and (
+                # Header with no text OR header is followed by space are valid conditions that sep is being used a header
+                len(stripped_line) == header_level or stripped_line[header_level] == " "
+            ):
+
+                # Pop out headers of lower or same level from the stack
+                while header_stack and header_stack[-1]["level"] >= header_level:
+                    # We have encountered a new header at the same or higher level
+                    popped_header = header_stack.pop()
+                    # Clear the metadata for the popped header in initial_metadata
+                    if popped_header["level"] in initial_metadata:
+                        initial_metadata.pop(popped_header["level"])
+
+                # Push the current header to the stack
+                header = {
+                    "level": header_level,
+                    "data": stripped_line[header_level :].strip(),
+                }
+                header_stack.append(header)
+                # Update initial_metadata with the current header
+                initial_metadata[header_level] = header["data"]
+
+                # Add the previous line to the lines_with_metadata only if current_text is not empty
+                if current_text:
+                    lines_with_metadata.append(
+                        {
+                            "text": "\n".join(current_text),
+                            "metadata": current_metadata.copy(),
+                        }
+                    )
+                    current_text.clear()
+
+                break
+        else:
+            if stripped_line:
+                current_text.append(stripped_line)
+            elif current_text:
+                lines_with_metadata.append(
+                    {
+                        "text": "\n".join(current_text),
+                        "metadata": current_metadata.copy(),
+                    }
+                )
+                current_text.clear()
+
+        current_metadata = initial_metadata.copy()
+
+    if current_text:
+        lines_with_metadata.append(
+            {"text": "\n".join(current_text), "metadata": current_metadata}
+        )
+
+    chunks = []
+    for line_id, line in enumerate(lines_with_metadata):
+        # Header path is "Header X" / "Sub Header Y" / "Sub Sub Header Z"
+        header_path = " / ".join(['"{}"'.format(header) for header in line["metadata"].values()])
+        chunks.append({"file": filename, "text": line["text"], "id": line_id + 1, "metadata": header_path, "error_message": ""})
```

Comment:
See PR comment, I think we should return the headers in a more structured way, i.e. a list. And possibly escape special characters ?